File: api/sdks/typescript/src/clients/jsonrpc.ts
/**
 * Shared JSON-RPC 2.0 envelope types for the MCP and Unix-socket clients,
 * matching the server's shared envelope in pkg/server/jsonrpc.
 */

import {
  ObjectStoreError,
  ObjectNotFoundError,
  AuthenticationError,
  AuthorizationError,
  ValidationError,
  AlreadyExistsError,
  RateLimitError,
  ServerError,
} from '../errors';

/** A JSON-RPC 2.0 request envelope. */
export interface JsonRpcRequest<P = unknown> {
  jsonrpc: '2.0';
  method: string;
  params: P;
  id: number;
}

/** A JSON-RPC 2.0 error object. */
export interface JsonRpcError {
  code: number;
  message: string;
  data?: unknown;
}

/** A JSON-RPC 2.0 response envelope. */
export interface JsonRpcResponse<R = unknown> {
  jsonrpc: '2.0';
  result?: R;
  error?: JsonRpcError;
  id: number;
}

/** Implementation-defined JSON-RPC error codes used by the objstore server. */
export const JSON_RPC_FORBIDDEN = -32001;
export const JSON_RPC_UNAUTHENTICATED = -32002;
export const JSON_RPC_NOT_FOUND = -32004;
export const JSON_RPC_ALREADY_EXISTS = -32005;
export const JSON_RPC_RATE_LIMITED = -32029;
/** Standard JSON-RPC 2.0 invalid-params error code. */
export const JSON_RPC_INVALID_PARAMS = -32602;

/**
 * errorFromJsonRpc converts a JSON-RPC error object from the objstore server
 * into the matching typed SDK error. The transport context (e.g. "Unix RPC",
 * "MCP") and the server's code and message are preserved in the error message.
 */
export function errorFromJsonRpc(error: JsonRpcError, context: string): ObjectStoreError {
  const detail = `${context} error (${error.code}): ${error.message}`;
  switch (error.code) {
    case JSON_RPC_NOT_FOUND:
      return new ObjectNotFoundError(detail);
    case JSON_RPC_UNAUTHENTICATED:
      return new AuthenticationError(detail);
    case JSON_RPC_FORBIDDEN:
      return new AuthorizationError(detail);
    case JSON_RPC_INVALID_PARAMS:
      return new ValidationError(detail);
    case JSON_RPC_ALREADY_EXISTS:
      return new AlreadyExistsError(detail);
    case JSON_RPC_RATE_LIMITED:
      return new RateLimitError(detail);
    default:
      return new ServerError(detail);
  }
}

File: api/sdks/python/objstore/_jsonrpc.py
"""Shared JSON-RPC 2.0 error mapping for the Unix and MCP transports.

Error codes mirror the server contract in pkg/server/jsonrpc: classification
is by code, never by message text.
"""

from typing import Dict

from objstore.exceptions import (
    AlreadyExistsError,
    AuthenticationError,
    AuthorizationError,
    ObjectNotFoundError,
    RateLimitError,
    ServerError,
    ValidationError,
)

# Implementation-defined server error codes (pkg/server/jsonrpc).
ERR_FORBIDDEN = -32001
ERR_UNAUTHENTICATED = -32002
ERR_NOT_FOUND = -32004
ERR_ALREADY_EXISTS = -32005
ERR_RATE_LIMITED = -32029

# Standard JSON-RPC 2.0 error codes.
ERR_PARSE_ERROR = -32700
ERR_INVALID_REQUEST = -32600
ERR_METHOD_NOT_FOUND = -32601
ERR_INVALID_PARAMS = -32602
ERR_INTERNAL = -32603


def raise_rpc_error(error: Dict) -> None:
    """Translate a JSON-RPC error object into an SDK exception.

    Args:
        error: JSON-RPC error dict with ``code`` and ``message`` fields

    Raises:
        ObjectStoreError: Converted exception
    """
    code: int = error.get("code", 0)
    message: str = error.get("message", "unknown error")

    if code == ERR_NOT_FOUND:
        raise ObjectNotFoundError(message)
    if code == ERR_UNAUTHENTICATED:
        raise AuthenticationError(message)
    if code == ERR_FORBIDDEN:
        raise AuthorizationError(message)
    if code == ERR_ALREADY_EXISTS:
        raise AlreadyExistsError(message)
    if code == ERR_RATE_LIMITED:
        raise RateLimitError(message)
    if code in (ERR_INVALID_PARAMS, ERR_INVALID_REQUEST):
        raise ValidationError(message)
    if code == ERR_INTERNAL:
        raise ServerError(message)
    raise ServerError(f"RPC error {code}: {message}")
